text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+
+This tool will stand as an analogue to `mycli` or `mysql-client` for users.
+
+## Motivation
+
+Currently `tikv-ctl` offers some ways to interact with a TiKV cluster, it
+focuses mostly on the *management* of a TiKV cluster instead of a *client* for
+the cluster.
+
+With the `tikv-client` crate users will find a way to interact with their TiKV
+cluster from their application. Some users will find it desirable to be able to
+inspect their cluster from a command line tool, or otherwise outside of their
+application.
+
+Building a simple command line tool to execute commands such as `get`, `set`,
+`delete`, and `scan` would mean users do not need to worry about creating such
+tools for themselves.
+
+It also provides us with a useful tool during the development cycle, as it will
+be very easy to issue one-off commands to TiKV.
+
+Further, we mentioned in the Rust Client RFC
+(https://github.com/tikv/rfcs/pull/7) we mentioned we wanted to accomplish this.
+
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
+functionality. It will utilize `clap` to provide a simple wrapper around the
+crate API.
+
+### Install
+
+We will build this tool in the `tikv-cli` repository. It will be published.
+Users will be able to install it with `cargo install tikv-cli`. (See
+*Alternatives* for another option.)
+
+### Configuration
+
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
+* The default options (connect to localhost, no TLS).
+
+The options are:
+
+```toml
+[connection]
+# A list of `pd` endpoints.
+pd_endpoints = [
+    "127.0.0.1:2379",
+    "127.0.0.2:2379",
+]
+
+[tls]
+# The path of the CA file for TLS.
+ca_path = "./rsa.ca"
+# The path of the certificate file for TLS.
+cert_path = "./rsa.cert"
+# The path of the key file for TLS.
+key_path = "./rsa.key"
+
+[output]
+# If the ouput JSON should be minified.
+minify = false
+# Level of logging to use. Available: error. warn, info, debug, trace
+logging = "info"
+```
+
+### Input/Output
+
+By default input is by positional arguments in `bash`/`sh` string format.
+
+By default the tool will output in prettyprinted JSON. For single entry results,
+it will return a string (including `"`), for key-value pairs it will use a JSON
+object `{ key: "", value: "" }`, for multiple results it will return a JSON
+array. Errors will use an `{ error: "" }` object. Opting out of prettyprint will
+be via the `--minify` flag.
+
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
+standard that has many tools (eg `jq`) to use for post processing if needed.
+
+All standard (non-logging) output will be on `stdout`.

```

Comment:
To clarify, "standard (non-logging) output" means the json stuff above? It would be good to mention that the total output on stdout will be valid JSON (will this be the case for stderr too?).

Reply:
Yes the intention is to make sure the output is fairly consistent on stdout so folks can use it in scripts.